ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/BeaconStateUtil.java

```diff
@@ -120,48 +118,46 @@ public static BeaconStateWithCache get_initial_beacon_state(
-      throws IllegalArgumentException {
-    UnsignedLong epoch = slot_to_epoch(slot);
-    UnsignedLong current_epoch = get_current_epoch(state);
-    UnsignedLong previous_epoch = get_previous_epoch(state);
-    UnsignedLong next_epoch = get_next_epoch(state);
+      BeaconState state, long slot, boolean registry_change) throws IllegalArgumentException {
+    long epoch = slot_to_epoch(slot);
+    long current_epoch = get_current_epoch(state);
+    long previous_epoch = get_previous_epoch(state);
+    long next_epoch = get_next_epoch(state);
 
     checkArgument(
-        previous_epoch.compareTo(epoch) <= 0 && epoch.compareTo(next_epoch) <= 0,
+        previous_epoch <= epoch && epoch <= next_epoch,
         "checkArgument threw and exception in get_crosslink_committees_at_slot()");
 
-    UnsignedLong committees_per_epoch = UnsignedLong.ZERO;
+    long committees_per_epoch = 0;
     Bytes32 seed = Bytes32.ZERO;
-    UnsignedLong shuffling_epoch = UnsignedLong.ZERO;
-    UnsignedLong shuffling_start_shard = UnsignedLong.ZERO;
+    long shuffling_epoch = 0;
+    long shuffling_start_shard = 0;
 
-    if (epoch.compareTo(current_epoch) == 0) {
+    if (epoch == current_epoch) {
       committees_per_epoch = get_current_epoch_committee_count(state);
       seed = state.getCurrent_shuffling_seed();
       shuffling_epoch = state.getCurrent_shuffling_epoch();
       shuffling_start_shard = state.getCurrent_shuffling_start_shard();
 
-    } else if (epoch.compareTo(previous_epoch) == 0) {
+    } else if (epoch == previous_epoch) {
       committees_per_epoch = get_previous_epoch_committee_count(state);
       seed = state.getPrevious_shuffling_seed();
       shuffling_epoch = state.getPrevious_shuffling_epoch();
       shuffling_start_shard = state.getPrevious_shuffling_start_shard();
 
-    } else if (epoch.compareTo(next_epoch) == 0) {
-      UnsignedLong current_committees_per_epoch = get_current_epoch_committee_count(state);
+    } else if (epoch == next_epoch) {
+
+      long current_committees_per_epoch = get_current_epoch_committee_count(state);
       committees_per_epoch = get_next_epoch_committee_count(state);
       shuffling_epoch = next_epoch;
-      UnsignedLong epochs_since_last_registry_update =
-          current_epoch.minus(state.getValidator_registry_update_epoch());
+      long epochs_since_last_registry_update =
+          current_epoch - state.getValidator_registry_update_epoch();
       if (registry_change) {
         seed = generate_seed(state, next_epoch);
         shuffling_start_shard =
-            state
-                .getCurrent_shuffling_start_shard()
-                .plus(current_committees_per_epoch)
-                .mod(UnsignedLong.valueOf(Constants.SHARD_COUNT));
-      } else if (epochs_since_last_registry_update.compareTo(UnsignedLong.ONE) > 0
+            state.getCurrent_shuffling_start_shard()
+                + current_committees_per_epoch % Constants.SHARD_COUNT;
```

Comment:
There is still an order of operations glitch here. Should be something like `(state.getCurrent_shuffling_start_shard()
                + current_committees_per_epoch) % Constants.SHARD_COUNT;`